src/nsrt_learning/strips_learning/gen_to_spec_learner.py

```diff
@@ -229,119 +277,187 @@ def _backchain_one_pass(
     def _finish_learning(
         self, param_opt_to_nec_pnads: Dict[ParameterizedOption,
                                            List[PartialNSRTAndDatastore]]
-    ) -> List[PartialNSRTAndDatastore]:
+    ) -> None:
         """Given the current PNADs where add effects and preconditions are
-        correct, learn the remaining components."""
-        # Make a list of all final PNADs. Note
-        # that these final PNADs only come from the
-        # param_opt_to_nec_pnads dict, since we can be assured
-        # that our backchaining process ensured that the
-        # PNADs in this dict cover all of the data!
-        nec_pnads = []
+        correct, learn the remaining components such that the resulting PNADs
+        are guaranteed to be harmless w.r.t the dataset.
+
+        Note that this may require spawning new PNADs with keep effects.
+        """
+        # Go thru all PNADs in the param_opt_to_nec_pnads dict and induce
+        # delete effects and side predicates.
+        for option, nec_pnad_list in sorted(param_opt_to_nec_pnads.items(),
+                                            key=str):
+            pnads_with_keep_effects = set()
+            for pnad in nec_pnad_list:
+                self._compute_pnad_delete_effects(pnad)
+                self._compute_pnad_side_predicates(pnad)
+                pnads_with_keep_effects |= self._get_pnads_with_keep_effects(
+                    pnad)
+            param_opt_to_nec_pnads[option].extend(
+                list(pnads_with_keep_effects))
+
+    def _reset_all_segment_add_effs(self) -> None:
+        """Reset all segment's necessary_add_effects to None."""
+        for ll_traj, seg_traj in zip(self._trajectories,
+                                     self._segmented_trajs):
+            if not ll_traj.is_demo:
+                continue
+            atoms_seq = utils.segment_trajectory_to_atoms_sequence(seg_traj)
```

Comment:
why do you need to compute the `atoms_seq` if you only need to iterate over its length? i don't understand why you can't just do: `for segment in seg_traj: segment.necessary_add_effects = None`

Reply:
is there some segment in each trajectory that you're purposefully _not_ resetting? if so, clarify that in a comment